File: src/main/scala/it/unibo/pps/duckgame/utils/GameUtils.scala
package it.unibo.pps.duckgame.utils

import it.unibo.pps.duckgame.controller.GameReader
import it.unibo.pps.duckgame.model.Player
import it.unibo.pps.duckgame.model.cell.specialCell.SpecialCell
import it.unibo.pps.duckgame.utils.PrologGameUtils.PrologEngine.PrologEngine
import it.unibo.pps.duckgame.utils.PrologGameUtils.PrologTheory
import it.unibo.pps.duckgame.utils.PrologGameUtils.PrologTheory.given
import it.unibo.pps.duckgame.utils.resources.PrologResources

import scala.language.implicitConversions
import scala.util.Random

object GameUtils:

  val CELLS_IN_SIDE = 7
  private val pathTheory = PrologResources.GAMEUTILS_PROLOG.path
  private val prolog: PrologEngine = PrologEngine(pathTheory)

  /** Method that returns the new position of current player If the position is
    * bigger than the gameboard's size, it will return back from the last cell
    *
    * @param sum
    *   result of the dices
    * @param position
    *   player's position to be updated
    * @return
    *   new player's position
    */
  def addSumToPosition(sum: Int)(position: Int): Int = sum + position match
    case result if result >= GameReader.gameBoard.size =>
      val updatedPosition = 63 - Math.abs(63 - result)
      updatedPosition
    case result => result

  /** Shuffles the order of the players in the list
    *
    * @param players
    *   A list og Players objects
    * @return
    *   A new list of Players objects with order randomized
    */
  def mixPlayers(players: List[Player]): List[Player] =
    Random shuffle players

  /** Returns the cell's coordinates given player's position First int is column
    * index, the second is the row
    *
    * @param position
    *   player's position
    * @return
    *   coordinates in (col, row) of the cell representing player's position
    */
  def getCoordinateFromPosition(position: Int): (Int, Int) =
    prolog.getFreeSlotInCell(position)(CELLS_IN_SIDE)

  /** Returns the coordinate of the free slot inside a grid of gridSize
    * dimensions
    * @param n
    *   The number of the slot in grid which coordinates are to be returned.
    * @param gridSize
    *   The dimensions of the grid (default 3 x 2)
    * @return
    *   The coordinates of the nth slot.
    */
  def getNthSlotFromCell(n: Int)(gridSize: (Int, Int)): (Int, Int) =
    prolog.getCellInGrid(n)(gridSize._1)(gridSize._2)

  /** Finds a SpecialCell object from the game board that corresponds to the
    * current player's position.
    *
    * @return
    *   An Option[SpecialCell] containing the SpecialCell if found, or None if
    *   not found.
    */
  def getSpecialCellFromPlayerPosition: Option[SpecialCell] =
    GameReader.gameBoard.specialCells.find(
      _.number == GameReader.currentPlayer.actualPosition
    )

File: src/main/resources/prolog/GameUtils.pl
% Return the coordinates of the N-th cell in a grid of given rows and columns
% N = Nth cell, Y = size of columns in grid, X = size of rows in grid
% Y1 = column in given position, X1 = row in given position

getCellInGrid(N, Y, X, Y1, X1):- R is (N mod Y), R = 0 -> Y1 is Y -1, X1 is N//Y - 1;
Y1 is ((N mod Y)-1), X1 is N//Y.


% Return the coordinates of a grid of cell given player's position
% P = position, C = CELLS IN SIDE (default 7)
% Y = column of N cell, X = row of N cell

getCoordinateFromPosition(P,C,Y,X):- P < C -> Y is P, X is 0;
P < C*2 -> Y is C, X is P - C;
P < C*3 -> Y is (C * 3) - P, X is C;
P < (C*4) - 1 -> Y is 0, X is (C*4)-P;
P < (C*5) - 2 -> Y is P - ((C*4) - 1) , X is 1;
P < (C*5) + 3 -> Y is C - 1, X is P -((C + 1)*4);
P =< C*6 -> Y is ((C*6)+2) - P, X is 6;
P < (C**2) - 2 -> Y is 1, X is ((C**2)-P);
P < (C**2)+2 -> Y is P-((C**2)-3), X is 2;
P < (C*8) -2 -> Y is 5 , X is P-(C**2);
P =< C*8 -> Y is ((C*8)+3)-P , X is 5;
P < (C*8) + 3 -> Y is 2 , X is ((C*9)-1)-P;
P < (C*9) - 2 -> Y is P-((C*8)+1) , X is 3;
P < C*9 -> Y is 4 , X is P-((C*8)+2);
Y is 3, X is 4.